tests/subprocesstest.py

```diff
@@ -0,0 +1,118 @@
+cases marked with @run_in_subprocess in a separate python interpreter.
+"""
+import os
+import subprocess
+import sys
+import unittest
+
+
+SUBPROC_TEST_ATTR = '_subproc_test'
+SUBPROC_ENV_VAR = '_SP_TEST'
+
+
+def run_in_subprocess(obj):
+    """
+    Marks a test case that is to be run in its own 'clean' interpreter instance.
+
+    When applied to a TestCase class, each method will be run in a separate
+    interpreter instance.
+
+    Usage on a class::
+
+        from tests.subprocesstest import SubprocessTestCase, run_in_subprocess
+
+        @run_in_subprocess
+        class PatchTests(SubprocessTestCase):
+            # will be run in new interpreter
+            def test_patch_before_import(self):
+                patch()
+                import module
+
+            # will be run in new interpreter as well
+            def test_patch_after_import(self):
+                import module
+                patch()
+
+
+    Usage on a test method::
+
+        class OtherTests(SubprocessTestCase):
+            @run_in_subprocess
+            def test_case(self):
+                pass
+
+
+    :param obj: method or class to run in a separate python interpreter.
+    :return:
+    """
+    setattr(obj, SUBPROC_TEST_ATTR, True)
+    return obj
+
+
+class SubprocessTestCase(unittest.TestCase):
+    def _full_method_name(self):
+        test = getattr(self, self._testMethodName)
+        modpath = test.__module__
+        clsname = self.__class__.__name__
+        testname = test.__name__
+        testcase_name = '{}.{}.{}'.format(modpath, clsname, testname)
+        return testcase_name
+
+    def _run_test_in_subprocess(self, result):
+        full_testcase_name = self._full_method_name()
+
+        sp_test_env = os.environ.copy()
+        sp_test_env[SUBPROC_ENV_VAR] = 'True'
+        sp_test_cmd = ['python', '-m', 'unittest', full_testcase_name]
+        sp = subprocess.Popen(
+            sp_test_cmd,
+            stdout=subprocess.PIPE,
+            stderr=subprocess.PIPE,
+            env=sp_test_env,
+        )
+        _, stderr = sp.communicate()
+
+        if sp.returncode:
+            try:
+                cmdf = ' '.join(sp_test_cmd)
+                raise Exception('Subprocess Test "{}" Failed'.format(cmdf))
+            except Exception:
+                exc_info = sys.exc_info()
+            sys.stderr.write(stderr)
+            result.addFailure(self, exc_info)
+        else:
+            result.addSuccess(self)
+
+    def _in_subprocess(self):
+        """Determines if the test is being run in a subprocess.
+
+        This is done by checking for an environment variable that we call the
+        subprocess test with.
+
+        :return: whether the test is a subprocess test
+        """
+        return bool(os.getenv(SUBPROC_ENV_VAR, False))
+
+    def _is_subprocess_test(self):
+        if hasattr(self, SUBPROC_TEST_ATTR):
+            return True
+
+        test = getattr(self, getattr(self, '_testMethodName'))
```

Comment:
You can make this more safe with `test = getattr(self, getattr(self, '_testMethodName', object()), None)`

Otherwise you might as well just do `test = getattr(self, self._testMethodName)`, since `self._testMethodName` is basically just `getattr(self, '_testMethodName')` anyways. If we don't care about attribute errors being raised here, then this is more clear and easy to follow.